src/coreclr/vm/riscv64/stubs.cpp

```diff
@@ -1307,151 +1361,370 @@ void StubLinkerCPU::EmitMovReg(IntReg Xd, IntReg Xm)
 void StubLinkerCPU::EmitMovReg(FloatReg dest, FloatReg source)
 {
     Emit32(RTypeInstr(0x53, 0, 0x11, dest, source, source));  // fsgnj.d
+    LOG((LF_STUBS, LL_EVERYTHING, "fmv %s, %s\n", fpRegAbiNames[dest], fpRegAbiNames[source]));
 }
 
-void StubLinkerCPU::EmitSubImm(IntReg Xd, IntReg Xn, unsigned int value)
+void StubLinkerCPU::EmitSubImm(IntReg Xd, IntReg Xn, int value)
 {
-    _ASSERTE(value <= 0x800);
-    EmitAddImm(Xd, Xn, ~value + 0x1);
+    EmitAddImm(Xd, Xn, -value);
 }
-void StubLinkerCPU::EmitAddImm(IntReg Xd, IntReg Xn, unsigned int value)
+void StubLinkerCPU::EmitAddImm(IntReg Xd, IntReg Xn, int value)
 {
     Emit32(ITypeInstr(0x13, 0, Xd, Xn, value));  // addi
+    if (value)
+        LOG((LF_STUBS, LL_EVERYTHING, "addi %s, %s, %i\n", intRegAbiNames[Xd], intRegAbiNames[Xn], value));
+    else
+        LOG((LF_STUBS, LL_EVERYTHING, "mv %s, %s\n", intRegAbiNames[Xd], intRegAbiNames[Xn]));
 }
 void StubLinkerCPU::EmitSllImm(IntReg Xd, IntReg Xn, unsigned int value)
 {
     _ASSERTE(!(value >> 6));
     Emit32(ITypeInstr(0x13, 0x1, Xd, Xn, value));  // slli
+    LOG((LF_STUBS, LL_EVERYTHING, "slli %s, %s, %u\n", intRegAbiNames[Xd], intRegAbiNames[Xn], value));
 }
 void StubLinkerCPU::EmitLuImm(IntReg Xd, unsigned int value)
 {
     _ASSERTE(value <= 0xFFFFF);
     Emit32((DWORD)(0x00000037 | (value << 12) | (Xd << 7))); // lui Xd, value
+    LOG((LF_STUBS, LL_EVERYTHING, "lui %s, %u\n", intRegAbiNames[Xd], value));
 }
 
 void StubLinkerCPU::Init()
 {
     new (gBranchIF) BranchInstructionFormat();
 }
 
+static bool InRegister(UINT16 ofs)
```

Comment:
Would it better to depend on the IL stub for more of the complex signatures (e.g. anything that involves floating point)?

This is 500 lines hand-written assembly emitter. Complicated hand-written assembly emitter like this are endless source of bugs. We have been trying to get rid of them where possible. 

If you think it is worth it to have this hand-written assembly emitter, could you please ask somebody from your team to review and signoff on it?

Reply:
Ideally, yes, but then we'd have to strip down the signatures of unnecessary information not to generate duplicate IL stubs, review the generated asm, and so on. I think we can postpone that to when we start optimizing RISC-V (2025 and on). There's a few more things to do to move away from generating stubs in stub linker even for simple shuffle thunks -- there aren't that many and we don't need full GenerateShuffleArray to know which one to use so I guess these could be pre-compiled.

For now, the hand-generated assembly itself isn't that complicated and it's already debugged, just the code is comment- and assert-heavy so that all the assumptions are explicit.

@dotnet/samsung could you review and sign-off please?